src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -672,69 +710,48 @@ internal PSMemberInfoInternalCollection<PSMemberInfo> InstanceMembers
             set => _instanceMembers = value;
         }
 
-        private PSMemberInfoInternalCollection<PSMemberInfo> _instanceMembers;
+        /// <summary>
+        /// This is the adapter that will depend on the type of baseObject.
+        /// </summary>
+        internal Adapter InternalAdapter => InternalAdapterSet.OriginalAdapter;
 
-        internal static bool HasInstanceMembers(object obj, out PSMemberInfoInternalCollection<PSMemberInfo> instanceMembers)
+        /// <summary>
+        /// This is the adapter that is used to resolve the base dotnet members for an
+        /// adapted object. If an object is not adapted, this will be null.
+        /// </summary>
+        /// <remarks>
+        /// If an object is not adapted, InternalAdapter will use the dotnet adapter.
+        /// So there is no point falling back to the same dotnet adapter.
+        ///
+        /// If an object is adapted, this adapter will be used to resolve the dotnet
+        /// members.
+        /// </remarks>
+        internal Adapter InternalBaseDotNetAdapter => InternalAdapterSet.DotNetAdapter;
+
+        /// <summary>
+        /// This is the adapter set that will contain the adapter of the baseObject
+        /// and the ultimate .net member lookup adapter.
+        /// See <see cref="PSObject.AdapterSet"/> for explanation.
+        /// </summary>
+        private AdapterSet InternalAdapterSet
         {
-            if (obj is PSObject psobj)
+            get
             {
-                lock (psobj)
+                if (_adapterSet == null)
                 {
-                    if (psobj._instanceMembers == null)
+                    lock (_lockObject)
```

Comment:
We could use Lazy

Reply:
That would add 24 byte of object header for Lazy. I'm trying to go in the opposite direction: Making PSObject leaner.

Reply:
Perhaps replacing lock with Interlocked.CompareExchange (Lazy) in PSObject could speed up the operation ~2x. In the file such lock is used in 9 places. For some cmdlets (Select-String, Import-Csv ...) it could be very noticeable.

It would be good to weight performance vs memory. 
Not for the PR.

Another thought. Perhaps there is scenarios in which we could initialize some properties in constructor without lock.